text/0000-socket-timeouts.md

```diff
@@ -0,0 +1,142 @@
+- Feature Name: socket_timeouts
+- Start Date: 2015-04-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add sockopt-style timeouts to `std::net` types.
+
+# Motivation
+
+Currently, operations on various socket types in `std::net` block
+indefinitely (i.e., until the connection is closed or data is
+transferred). But there are many contexts in which timing out a
+blocking call is important.
+
+The [goal of the current IO system][io-reform] is to gradually expose
+cross-platform, blocking APIs for IO, especially APIs that directly
+correspond to the underlying system APIs. Sockets are widely available
+with nearly identical system APIs across the platforms Rust targets,
+and this includes support for timeouts via [sockopts][sockopt].
+
+So timeouts are well-motivated and well-suited to `std::net`.
+
+# Detailed design
+
+The proposal is to *directly expose* the timeout functionality
+provided by [`setsockopt`][sockopt], in much the same way we currently
+expose functionality like `set_nodelay`:
+
+```rust
+impl TcpStream {
+    pub fn set_read_timeout(&self, dur: Option<Duration>) -> io::Result<()> { ... }
```

Comment:
The underlying OS API is just `setsockopt` which is already threadsafe and we've specifically avoided adding a `&mut self` barrier on `Tcp{Stream,Listener}` for example with `Read for &T` or having `accept` take `&self`, even though they're all kinda though of as "mutating" the underlying stream.

In other words, the `&mut self` isn't necessary, so the primitives don't require it.